File: workspace/micke/bytesview.py
# Copyright (C) 2019-2021 Data Ductus AB
#
# Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
#
# 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#

# Python version of bytesview

##################################################
### Bytesview

class Bytesview():

    _bytes : bytes = None
    _pre = None
    _start : int = 0
    _end : int = 0
    _eof : bool = False
    
    def __init__(self, b, n=None, start=None, pre=None, closed=False):
        self._bytes = b
        self._pre = pre 
        self._start = 0 if start is None else start
        if n is not None:
            self._end = self._start + n
        else:
            self._end = len(b)-self._start
        self._eof = closed

    ### analysis of Bytesview objects
    
    def n(self):
        return self._end - self._start + (self._pre.n() if self._pre is not None else 0)

    def to_bytes(self):
        these = self._bytes[self._start:self._end]
        if self._pre is None:
            return these
        else:
            return self._pre.to_bytes() + these

    ### de-lousing

    def show(self):
        if self._pre is not None:
            self._pre.show()
        print(self._bytes, "[{}:{}]".format(self._start,self._end))
        if self._eof:
            print("_eof = {}".format(self._eof))
            
    ### consuming data

    def _consume(self, amount):
        if self._pre is not None:
            (pre, remaining) = self._pre._consume(amount)
        else:
            pre = None
            remaining = amount
            
        thislen = self._end - self._start
        if remaining < thislen:
            newthis = Bytesview(self._bytes, n=thislen-remaining, start=self._start+remaining, pre=pre)
            newremaining = 0
        else:
            newthis = pre
            newremaining = remaining - thislen
        
        return(newthis,newremaining)
        
    def consume(self, amount):
        # assert(self.n()>=amount)
        (newview, remaining) = self._consume(amount)
        if remaining > 0:
            # raise IndexError
            return None

        if newview is not None:
            return newview
        else:
            return Bytesview(b'')

#    def _initialbytes(self, amount=a, end=-1, endskip=0):
#        
#
#    def initialbytes(self, end=-1, endskip=0):
#        # remove data at end of bytesview.
#        #
#        # end == -1 implies trimming from eof (so error if -1 and not closed)
#        # end >= 0 implies end-endskip is the index where returned bytesview object ends.
#        # remove endskip bytes before end.
#        if end == -1 and not self._eof:
#            raise ValueError("Trim(end=-1) on non-closed Bytesview")
#
#        if self._pre is None:
#            
#        thislen = self._end - self._start
#        thislen -= 
#        if self._pre is None:
#            newstartskip = startskip
#            newend = end
#            newenskip = endskip
#            return Bytesview(self._bytes, start=self._start+ 
#        else:
#            
#        if start is None:
#            if end is None or end == -1:
#                these = self._bytes[self._start:self._end]
#            else:
#                these = self._bytes[self._start:self._start+end]
#        else:
#            if end is None or end == -1:
#                these = self._bytes[self._start+start:self._end]
#            else:
#                these = self._bytes[self._start+start:self._start+end]
#        if self._pre is None:
#            return these
#        else:
#            return self._pre.to_bytes() + these

    ### reading out data from a Bytesview object (creating a new)
    ###
    ### Raises IncompleteReadError if Bytesview object does not contain
    ### enough data to fulfill request.

    def _read(self, n):
       # n is always >= 0
       if self._pre is not None:
           (remaining, initial) = self._pre._read(n)
       else:
           remaining = n
           initial = None
 
       thislen = self._end - self._start
       if remaining == 0:
           return (remaining,initial)
       elif remaining >= thislen:
           return (remaining-thislen, self)
       else:  # 0 < remaining < thislen
           return (0, Bytesview(self._bytes, start=self._start, n=remaining, pre=initial))
                             

    def read(self, n=-1):                  # (int) -> (bytes, Bytesview) 
       if n == -1:
           if not self._eof:
               raise IncompleteReadError("read(-1) on non-closed Bytesview")
               return None
           else:
               return self
       else:
           # deliver n bytes, if there are n bytes.
           # if not, deliver available bytes if closed, error if not closed
           thislen = self._end - self._start
           if self._pre is not None:
               (remaining, initial) = self._pre._read(n)
           else:
               remaining = n
               initial = None

           if remaining == 0:
               return initial if initial is not None else Bytesview(b'')
           elif remaining < thislen:
               return Bytesview(self._bytes, n=remaining, start=self._start, pre=initial)
           elif remaining > thislen:
               if self._eof:
                   return self
               else:
                   #raise IncompleteReadError("read(n) on non-closed Bytesview with less than n bytes")
                   return None
           else:
               # remaining == thislen
               return self
        
    def readline(self):
        return self.readuntil (separator=b'\n')

    def _readexactly(self, n):
        thislen = self._end - self._start
        if self._pre:
            (initial, remains) = self._pre._readexactly(n)
        else:
            (initial, remains) = (None, n)

        if remains == 0:
            return (initial, remains)
        elif remains >= thislen:
            return (self, remains-thislen)
        else:
            # remains < thislen
            return (Bytesview(self._bytes, start=self._start, n=remains, pre=initial), 0)
        
    def readexactly(self, n):
        (initial, remains) = self._readexactly(n)
        if remains > 0:
            #raise IncompleteReadError("readexactly(): Too few bytes available")
            return None
        if initial is None:
            initial = Bytesview(b'')
        return initial 
    
    def _readuntil(self, separator, seplen):
        # len(last) < len(separator)
        if self._pre is not None:
            (last, initial, found) = self._pre._readuntil(separator, seplen)
        else:
            (last, initial, found) = (b'', None, False)

        # last is potential beginning of separator at end of previous Bytesview(s)

        if found:
            return (b'',initial, found)
        else:
            buf = last+self._bytes[self._start:self._start+seplen-1]
            idx = buf.find(separator)
            if idx >= 0:   # found it!
                # end of separator at ix+seplen-len(last)
                #print (buf, len(buf), last, len(last), sep, idx)
                return (b'', Bytesview(self._bytes, n=idx+seplen-len(last),
                                       start=self._start,
                                       pre=self._pre), True)
            idx = self._bytes.find(separator, self._start)
            if idx >= 0:   # found it!
                return (b'', Bytesview(self._bytes, n=idx-self._start+seplen, start=self._start, pre=self._pre), True)
            thislen = self._end-self._start
            if thislen >= seplen-1:
                last = self._bytes[self._end-(seplen-1):self._end]
            else:
                last = last[thislen-(seplen-1):] + self._bytes[self._start:self._end]
            return (last, self, False)
                    
    def readuntil(self, separator=b'\n'):
        seplen = len(separator)
        (_, initial, found) = self._readuntil(separator, seplen)
        if found:
            return initial if initial is not None else Bytesview(b'')
        else:
            if self._eof:
                #raise IncompleteReadError("Separator not found and Bytesview closed")
                return None
            else:
                #raise IncompleteReadError("Separator not found")
                return None
    
    def at_eof(self):
        return self._eof and self._pre is None and self._start == self._end
    
    ### feeding data & closing
    
    def append(self, b, n=None):
        if self._eof:
            #raise ValueError("append() on closed Bytesview")
            return None
        if self._start < self._end:
            return Bytesview(b, n=n, pre=self)
        else:
            return Bytesview(b, n=n)

    def write(self, b, n=None):  	# same as append(self, b, n=None)
        if self._eof:
            #raise ValueError("write() on closed Bytesview")
            return None
        return self.append(b, n)
        
    def writelines(self,data):
        if self._eof:
            #raise ValueError("writelines() on closed Bytesview")
            return None
        for b in data:
            self.append(b)
        
    def close(self):
        if self._eof:
            return self
        return Bytesview(self._bytes, n=self._end-self._start, start=self._start, pre=self._pre, closed=True)

##################################################
### Unit tests

def report(bv, name):
    print (name+" = ", bv)
    print (name+".n() = ", bv.n())
#    print (name+"limits() = ", bv.limits())
    print (name+".to_bytes() = ", bv.to_bytes())
    print (name+".show():")
    bv.show()

by = b'0123456789abcdef'

#l = len(by)
#print("by = ",by)
#print("l = ", l)
#
#bv1 = Bytesview(by)
#bv2 = Bytesview(by, n=16)
#                 
#report(bv1, "bv1")
#report(bv2, "bv2")

bv3 = bv2.append(by, n=16)
#report(bv2, "bv2")
#report(bv3, "bv3")

bv4 = bv3.append(by)
#report(bv4,"bv4")

#bv = bv4
#n=0
#while True: 
#   res = bv.consume(n)
#   print("")
#   res.show()
#   #print(res.to_bytes())
#   n += 1   

###########
### read()
#
#bv = bv4
#n=0
#while True: 
#   res = bv.read(n)
#   print("")
#   res.show()
#   #print(res.to_bytes())
#   n += 1   

#print("read() test")

#bv = bv4
#n=0
#while n < 60: 
#   res = bv.read(n)
#   print(res.to_bytes())
#   n += 1

#bv = bv4.close()
#print("bv length: ", bv.n())
#
#n=0
#while n < 60: 
#   res = bv.read(n)
#   print(res.to_bytes())
#   n += 1

#bv = bv4.close()
#print("bv length: ", bv.n())
#
#n=0
#while n < 60: 
#   res = bv.read(n)
#   print(res.to_bytes())
#   res.show()
#   n += 1

###########
### readuntil()
#

#separators = by+b'x'+by+b'y'+by+b'z'+by+b'w'
#bv = Bytesview(by+b'x').append(by+b'y').append(by+b'z').append(by)
#bv.close()
#print("bv length: ", bv.n())
#
#n=0
#while n < 4*16+3+5:
#    sep = separators[n:n+7]
#    res = bv.readuntil(separator=sep)
#    print(sep, " ", res.to_bytes())
#    res.show()
#    n += 1

def bv_list(bv=None):
    bvl = bv if bv is not None else Bytesview(b'')
    n = 0
    while n < 16:
        bvl = bvl.append(by[n:n+1])
        n += 1
    return bvl

bvl = bv_list().append(b'x')
bvl = bv_list(bvl).append(b'y')
bvl = bv_list(bvl).append(b'z')
bvl = bv_list(bvl)

print ("bvl: ", bvl.to_bytes())
bvl.show()

#bv.close()
print("bv length: ", bvl.n())

separators = by+b'x'+by+b'y'+by+b'z'+by+b'w'
n=0
while n < 4*16+3+5:
    sep = separators[n:n+7]
    res = bvl.readuntil(separator=sep)
    print(sep, " ", res.to_bytes())
    res.show()
    n += 1   

###########
### readexactly()
#

#bv = Bytesview(by+b'x').append(by+b'y').append(by+b'z').append(by)
#bv.close()
#print("bv length: ", bv.n())
#
#n=0
#while n < 4*16+3+5:
#    res = bv.readexactly(n)
#    print(res.to_bytes())
#    #res.show()
#    n += 1
#

File: workspace/micke/bytesview.c
/*
 * Copyright (C) 2019-2021 Data Ductus AB
 *
 * Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/* bytesview.c */
    
#include <assert.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "bytesview.h"

bytesview_t const _empty = {
    ._pre = NULL,
    ._eof = 0,
    .buf = { ._start = 0, ._end = 0, ._bytes = (uint8_t *) ""}
};


act_bv_method_table_t *_bv_meth;

/* ### creation ### */

bytesview_t const *act_bv_init (const uint8_t *b, size_t n, size_t start, const bytesview_t *pre, int closed)
{
    size_t len;
    bytesview_t *bv = (bytesview_t *) malloc(sizeof(bytesview_t));

    bv->_pre = (bytesview_t *) pre;              /* disable warning: discards const qualifier */
    bv->_eof = closed;
    if (start >= 0) {
        bv->buf._start = start;
    } else {
        bv->buf._start = 0;
    }
    if (n >= 0) {
        bv->buf._end = bv->buf._start+n;
    } else {
        len = strlen((char *) b);
        bv->buf._end = len - bv->buf._start;
    }
    bv->buf._bytes = (uint8_t *) b;              /* disable warning: discards const qualifier */
      
    return bv;
}

bytesview_t const *act_bv_empty()
{
    return &_empty;
}

bytesview_t const *act_bv_fromstr (char const *s)
{
    return act_bv_init((uint8_t const *)s, strlen(s), 0, NULL, 0);
}

bytesview_t const *act_bv_frombuf(bytesview_data_buffer_t const *buf)
{
    bytesview_t *bv = (bytesview_t *) malloc(sizeof(bytesview_t));

    bv->_pre = NULL;
    bv->_eof = 0;
    bv->buf._start = buf->_start;
    bv->buf._end = buf->_end;
    bv->buf._bytes = buf->_bytes;

    return bv;
}

/* ### analysis ### */

size_t act_bv_n(bytesview_t const *self)
{
    size_t prelen;

    if (self->_pre == NULL) {
        prelen = 0;
    } else {
        prelen = act_bv_n(self->_pre);
    }
    return prelen + self->buf._end - self->buf._start;
}

static uint8_t *act_bv_cpybuf(bytesview_t const *self, uint8_t *p)
{
    /* copy all bytes to *p */

    uint8_t *q;
    size_t thislen;
    
    if (self->_pre != NULL) {
        q = act_bv_cpybuf(self->_pre, p);
    } else {
        q = p;
    }
    thislen = self->buf._end - self->buf._start;
    memcpy(q, self->buf._bytes+self->buf._start, thislen);
    return q+thislen;
}

bytesview_data_buffer_t const *act_bv_tobuf(bytesview_t const *self)
{
    size_t n = act_bv_n(self);
    bytesview_data_buffer_t *p;

#if NDEBUG
#else
    uint8_t *q;
#endif

    if (self->buf._end - self->buf._start == n) {
        /* no more than one buffer */
        p = (bytesview_data_buffer_t *) &(self->buf);
    } else {
        /* several buffers, must concatenate */
        p = (bytesview_data_buffer_t *) malloc(sizeof(bytesview_data_buffer_t));
        p->_bytes = (uint8_t *) malloc (n);
#if NDEBUG
        act_bv_cpybuf(self, p->_bytes);
#else
        q  = act_bv_cpybuf(self, p->_bytes);
        assert(q-p->_bytes == n);
#endif		      
        p->_start = 0;
        p->_end = n;
    }
    return p;
}

char *act_bv_tostr(bytesview_t const *self)
{
    size_t n = act_bv_n(self);
    uint8_t *p, *q;
    
    p = (uint8_t *) malloc(n+1);  /* room for NUL character at end */
    q = act_bv_cpybuf(self, p);
    assert(q-p == n);
    *q = 0;    /* NUL character at end */
    return (char *) p;
}


int act_bv_at_eof (const bytesview_t *self)
{
    return self->_eof;
}

/* ### de-lousing */
void act_bv_showbuf(const bytesview_data_buffer_t *buf)
{
    size_t i;
    size_t thislen;
    uint8_t *p;

    thislen = buf->_end - buf->_start;
    p = buf->_bytes+buf->_start;
    
    printf("[%lu,%lu] b'", buf->_start, buf->_end);
    i = 0;
    while (i++ < thislen) {
        printf("%c", *p++);
    }
    printf("'\n");
}

void act_bv_show(const bytesview_t *self)
{
    if (self == NULL) return;

    act_bv_show(self->_pre);
    printf("_eof=%d - buf= ", self->_eof);
    act_bv_showbuf(&self->buf);
}

void raise_IncompleteReadError(char *msg)
{
    fprintf(stderr, "IncompleteReadError: %s\n", msg);
    // assert(0==1);
}
      
void raise_IndexError(char *msg)
{
    fprintf(stderr, "IndexError: %s\n", msg);
    // assert(0==1);
}

void raise_ValueError(char *msg)
{
    fprintf(stderr, "ValueError: %s\n", msg);
    // assert(0==1);
}

/* ### consuming data */

void act_bv__consume(bytesview_t const *self, int amount, bytesview_t const **prep, size_t *remainingp)
{
    bytesview_t const *pre;
    bytesview_t const *newthis;
    size_t remaining, newremaining;
    size_t thislen;
    
    if (self->_pre != NULL) {
        act_bv__consume(self->_pre, amount, &pre, &remaining);     
    } else {
        pre = NULL;
        remaining = amount;
    }
    thislen = self->buf._end - self->buf._start;
    if (remaining < thislen) {
        newthis = act_bv_init(self->buf._bytes, thislen-remaining, self->buf._start+remaining, pre, self->_eof);
        newremaining = 0;
    } else {
        newthis = pre;
        newremaining = remaining - thislen;
    }
    *prep = newthis;
    *remainingp = newremaining; 
    return;
}

bytesview_t const *act_bv_consume(bytesview_t const *self, size_t amount)
{
    bytesview_t const *newview;
    size_t remaining;

    act_bv__consume(self, amount, &newview, &remaining);
    if (remaining > 0) {
        raise_IndexError("Not enough data to consume");
        return NULL;
    }

    if (newview != NULL) {
        return newview;
    } else {
        return act_bv_empty();
    }
}


void act_bv__read(bytesview_t const *self, int n, int *remainingp, bytesview_t const **initialp)
{
    bytesview_t const *initial;
    int remaining;
    size_t thislen;

    assert(n>=0);

    if (self->_pre != NULL) {
        act_bv__read(self->_pre, n, &remaining, &initial);
    } else {
        remaining = n;
        initial = NULL;
    }
    thislen = self->buf._end-self->buf._start;
    if (remaining == 0) {
        *remainingp = remaining;
        *initialp = initial;
    } else if (remaining >= thislen) {
        *remainingp = remaining-thislen;
        *initialp = self;
    } else {
        /*  0 < remaining < thislen */
        *remainingp = 0;
        *initialp = act_bv_init(self->buf._bytes, remaining, self->buf._start, initial, 0);
    }
    return;
}

bytesview_t const *act_bv_read(bytesview_t const *self, size_t n)
{
    bytesview_t const *initial;
    int remaining;
    int thislen;
      
    if (n == -1) {
        if (!self->_eof) {
            raise_IncompleteReadError("read(-1) on non-closed Bytesview");
            return NULL;
         } else {
            return (bytesview_t *) self;
         }
    } else {
         /* deliver n bytes, if there are n bytes.
          * if not, deliver available bytes if closed, error if not closed
          */
        thislen = self->buf._end - self->buf._start;
        if (self->_pre != NULL) {
            act_bv__read(self->_pre, n, &remaining, &initial);
        } else {
            remaining = n;
            initial = NULL;
        }

        if (remaining == 0) {
            if (initial != NULL) {
                return initial;
            } else {
                return act_bv_empty();
            }
        } else if (remaining < thislen) {
            return act_bv_init(self->buf._bytes, remaining, self->buf._start, initial, 0);
        } else if (remaining > thislen) {
          if (self->_eof) return (bytesview_t *) self;
            else raise_IncompleteReadError("read(n) on non-closed Bytesview with less than n bytes");
            return NULL;
        } else {
            /* remaining == thislen */
          return (bytesview_t *) self;
        }
    } 
}

bytesview_t const *act_bv_readline(bytesview_t const *self)
{
    return act_bv_readuntil(self, "\n");
}

static void act_bv__readexactly(bytesview_t const *self, size_t n, bytesview_t const **initialp, size_t *remainsp)
{
    bytesview_t const *initial;
    size_t remains, thislen;

    thislen = self->buf._end - self->buf._start;
    if (self->_pre != NULL) {
        act_bv__readexactly(self->_pre, n, &initial, &remains);
    } else {
        initial = NULL;
        remains = n;
    }

    if (remains == 0) {
        *initialp = initial;
        *remainsp = remains;
    } else if (remains >= thislen) {
        *initialp = (bytesview_t *) self;
        *remainsp = remains - thislen;
    } else {
        /* remains < thislen */
        *initialp = act_bv_init(self->buf._bytes, remains, self->buf._start, initial, 0);
        *remainsp = 0;
    }
}

bytesview_t const *act_bv_readexactly(const bytesview_t *self, size_t n) {
    bytesview_t const *initial;
    size_t remains;

    act_bv__readexactly(self, n, &initial, &remains);
    if (remains > 0) {
        raise_IncompleteReadError("readexactly(): Too few bytes available");
        return NULL;
    }

    if (initial == NULL) {
        initial = act_bv_empty();
    }
    
    return initial;
}

static size_t act_bv_min(size_t a, size_t b)
{
    if (a <= b) return a;
    else return b;
}

static int act_bv__readuntil(bytesview_t const *self, void const *separator, size_t seplen,
    bytesview_data_buffer_t *last,
    bytesview_t const ** initialp)
{
    bytesview_t const *initial;
    int found;
    uint8_t *m;
    size_t idx, lastlen, cpylen, thislen, keeplen;

    if (self->_pre != NULL) {
        found = act_bv__readuntil(self->_pre, separator, seplen, last, &initial);
    } else {
        /* last initialized in act_bv_readuntil() */
        initial = NULL;
        found = 0;
        //printf("-");	
    }
    // last is potential beginning of separator at end of previous buffer(s)

    if (found) {
        /* last made empty in recursive call */
        *initialp = initial;
        return found;
    }
    //printf("-\n");
    
    /* still looking - is separator straddling buffers? */
    lastlen = last->_end - last->_start;
    thislen = self->buf._end - self->buf._start;
    cpylen = act_bv_min(seplen-1, thislen);
    memcpy(last->_bytes+last->_end, self->buf._bytes+self->buf._start, cpylen);
    lastlen = last->_end += cpylen;
    /* now, last contains last part of previous plus first part of this */
    //printf("lastlen = %ld, cpylen = %ld\n", lastlen, cpylen);
    //printf("last: ");
    //act_bv_showbuf(last);
    if (lastlen >= seplen) {
        // might be straddling, check!
        m = (uint8_t *) strstr((const char *)last->_bytes+last->_start, separator);
        if (m != NULL) {
            found = 1;
	    idx = m - (last->_bytes+last->_start);
            //printf("straddling! idx=%ld cpylen=%ld last: [%ld, %ld]\n", idx, cpylen, last->_start, last->_end);
	    // end of separator at idx + seplen - len(last)
	    *initialp = act_bv_init(self->buf._bytes, cpylen-(last->_end-(idx+seplen)), self->buf._start, self->_pre, 0); 
	    // last->_start = last->_end = 0;
	    return found;
	}
    }
    
    /* Nope, so search this buffer */
    m = memmem(self->buf._bytes+self->buf._start, self->buf._end-self->buf._start, separator, seplen);
    if (m != NULL) {
	found = 1;
	idx = m - (self->buf._bytes + self->buf._start);
        //printf("inbuf [%ld, %ld]: idx == %ld\n", self->buf._start, self->buf._end, idx);
	*initialp = act_bv_init(self->buf._bytes, idx+seplen, self->buf._start, self->_pre, 0);
	return found;
    }
    
    /* Nope, so return last part of this buffer in case separator is straddling */
    found = 0;
    if (thislen >= seplen-1) {
	/* discard old contents of last */
	last->_start = 0;
	last->_end = seplen-1;
	memcpy(last->_bytes, self->buf._bytes+self->buf._end-(seplen-1), seplen-1);
	*(last->_bytes+last->_end) = 0;
    } else {
        /* entire this is in last, so just keep end of last */
        keeplen = act_bv_min(seplen-1, lastlen);
	if (keeplen < lastlen) {
	    // move end of last up front  
	    memcpy(last->_bytes, last->_bytes+last->_end-keeplen, keeplen);
	    last->_start = 0;
	    last->_end = keeplen;
 	    *(last->_bytes+last->_end) = 0;
	} else {
	    /* keep all of last */
	}
    }
    *initialp = self;
    return found;
}

bytesview_t const *act_bv_readuntil(bytesview_t const *self, void const *separator)
{
    bytesview_data_buffer_t buf, *last;
    size_t seplen;
    bytesview_t const *initial;
    int found;

    seplen = strlen(separator);
    last = &buf;
    last->_start = 0;
    last->_end = 0;
    last->_bytes = (uint8_t *) malloc (seplen + seplen -1);   // room for (seplen-1)*2 bytes + nul byte
    *(last->_bytes) = 0;
    
    found = act_bv__readuntil(self, separator, seplen, last, &initial);

    if (found) {
        if (initial != NULL) {
            return initial;
        } else {
            return act_bv_empty();
        }
    } else {
        if (self->_eof) {
            raise_IncompleteReadError("Separator not found and Bytesview closed");
            return NULL;
        } else {
            raise_IncompleteReadError("Separator not found");
            return NULL;
        }
    }
}

/* ### feeding data & closing */

bytesview_t const *act_bv_append(bytesview_t const *self, void const *b, size_t n)
{
    if (self->_eof) {
        raise_ValueError("append() on closed Bytesview");
	return NULL;
    }
    if (self->buf._start < self->buf._end) {
        return act_bv_init(b, n, 0, self, 0);
    } else {
        return act_bv_init(b, n, 0, NULL, 0);
    }
}

bytesview_t const *act_bv_write(bytesview_t const *self, void const *b, size_t n)
{
    if (self->_eof) {
        raise_ValueError("write() on closed Bytesview");
	return NULL;
    } 
    return act_bv_append(self, b, n);
}

bytesview_t const *act_bv_writelines(bytesview_t const *self, void const **data)
{
    if (self->_eof) {
        raise_ValueError("writelines() on closed Bytesview");
    }

    /* XXX */
    assert(0);
    return act_bv_empty();
}

bytesview_t const *act_bv_close (bytesview_t const *self)
{
    if (self->_eof) return (bytesview_t *) self;             /* disable warning: discards const qualifier */
    return act_bv_init(self->buf._bytes, self->buf._end-self->buf._start, self->buf._start, self->_pre, 1);
}


/* method table */

act_bv_method_table_t _act_bv_method_table = {
    .init = act_bv_init,
    .empty = act_bv_empty
};

void act_bv_class_init() {
    _bv_meth = & _act_bv_method_table;
}

